File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/ViewBandedDemo/ViewBandedDemoMain.pas
unit ViewBandedDemoMain;

interface

uses
  Windows, Messages, Classes, Graphics, Controls, Forms,
  Dialogs, cxGridCustomTableView, cxGridTableView, cxGridBandedTableView,
  cxGridDBBandedTableView, cxControls, cxGridCustomView, StdCtrls, Menus,
  cxGridLevel, cxGrid, ActnList, ImgList, cxLookAndFeels,
  ComCtrls, cxStyles, cxCustomData, cxGraphics, cxFilter,
  cxData, cxEdit, cxClasses, cxListBox, cxDataStorage, DB, cxDBData,
  cxImageComboBox, cxDBLookupComboBox, cxCalendar, cxLookAndFeelPainters,
  BaseForm, cxNavigator, cxGridCardView;

type
  TViewBandedDemoMainForm = class(TfmBaseForm)
    glItems: TcxGridLevel;
    cxGrid: TcxGrid;
    btvItems: TcxGridDBBandedTableView;
    btvItemsNAME: TcxGridDBBandedColumn;
    btvItemsTYPE: TcxGridDBBandedColumn;
    btvItemsPROJECTID: TcxGridDBBandedColumn;
    btvItemsPRIORITY: TcxGridDBBandedColumn;
    btvItemsSTATUS: TcxGridDBBandedColumn;
    btvItemsCREATEDDATE: TcxGridDBBandedColumn;
    btvItemsFIXEDDATE: TcxGridDBBandedColumn;
    btvItemsDESCRIPTION: TcxGridDBBandedColumn;
    btvItemsOwnerName: TcxGridDBBandedColumn;
    btvItemsCreatorName: TcxGridDBBandedColumn;
    miOptions: TMenuItem;
    ilMain: TImageList;
    btvItemsCreatorPhone: TcxGridDBBandedColumn;
    btvItemsOwnerPhone: TcxGridDBBandedColumn;
    miShowPreviewRow: TMenuItem;
    miBandMoving: TMenuItem;
    miBandSizing: TMenuItem;
    miDeleteBand: TMenuItem;
    miCreateBand: TMenuItem;
    miColumnsCustomization: TMenuItem;
    N1: TMenuItem;
    miShowNavigator: TMenuItem;
    procedure FormShow(Sender: TObject);
    procedure miShowPreviewRowClick(Sender: TObject);
    procedure miBandMovingClick(Sender: TObject);
    procedure miBandSizingClick(Sender: TObject);
    procedure miColumnsCustomizationClick(Sender: TObject);
    procedure miCreateBandClick(Sender: TObject);
    procedure miDeleteBandClick(Sender: TObject);
    procedure miShowNavigatorClick(Sender: TObject);
  private
    procedure DeleteBand(ABandCaption: String);
    function GetBandByCaption(ABandCaption: String): TcxGridBand;
  protected
    procedure UpdateButtons;
  end;

var
  ViewBandedDemoMainForm: TViewBandedDemoMainForm;

implementation

{$R *.dfm}

uses
  ViewBandedDemoData, ViewBandeDemoBands, AboutDemoForm;

procedure TViewBandedDemoMainForm.FormShow(Sender: TObject);
begin
  TcxGridDBBandedTableView(cxGrid.FocusedView).DataController.Groups.FullExpand;
end;

procedure TViewBandedDemoMainForm.miShowPreviewRowClick(Sender: TObject);
begin
  if GetMenuItemChecked(Sender) then
  begin
    //btvItemsDESCRIPTION.Visible := False;
    btvItems.Preview.Column := btvItemsDESCRIPTION;
    btvItems.Preview.Visible := True;
  end
  else
  begin
    btvItems.Preview.Visible := False;
    btvItems.Preview.Column := nil;
  end;
end;

procedure TViewBandedDemoMainForm.miBandMovingClick(Sender: TObject);
begin
  btvItems.OptionsCustomize.BandMoving := GetMenuItemChecked(Sender);
end;

procedure TViewBandedDemoMainForm.miBandSizingClick(Sender: TObject);
begin
  btvItems.OptionsCustomize.BandSizing := GetMenuItemChecked(Sender);
end;

procedure TViewBandedDemoMainForm.DeleteBand(ABandCaption: String);
begin
  GetBandByCaption(ABandCaption).Free;
end;

function TViewBandedDemoMainForm.GetBandByCaption(ABandCaption: String): TcxGridBand;
var
  i: Integer;
begin
  Result := nil;
  for i := 0 to btvItems.Bands.Count - 1 do
    if btvItems.Bands[i].Caption = ABandCaption then
    begin
      Result := btvItems.Bands[i];
      Exit;
    end;
end;

procedure TViewBandedDemoMainForm.UpdateButtons;
begin
  MenuItemSetEnabled('miDeleteBand', btvItems.Bands.Count <> 0);
  MenuItemSetEnabled('miShowPreviewRow', btvItems.Bands.Count <> 0);
end;

procedure TViewBandedDemoMainForm.miColumnsCustomizationClick(Sender: TObject);
begin
  btvItems.Controller.Customization := True;
end;

procedure TViewBandedDemoMainForm.miCreateBandClick(Sender: TObject);
var
  BandCaption: String;
  ABand: TcxGridBand;
  ABandViewInfo: TcxGridbandViewInfo;
begin
  if InputQuery('Create band', 'Specify a caption of the band', BandCaption) then
    if GetBandByCaption(BandCaption) <> nil then
      MessageDlg('Band with this caption already exists', mtWarning, [mbOK], 0)
    else
    begin
      ABand := btvItems.Bands.Add;
      ABand.Caption := BandCaption;
      ABandViewInfo :=
        btvItems.ViewInfo.HeaderViewInfo.BandsViewInfo[ABand.VisibleIndex];
      btvItems.Controller.LeftPos := ABandViewInfo.Bounds.Right;
    end;
  UpdateButtons;
end;

procedure TViewBandedDemoMainForm.miDeleteBandClick(Sender: TObject);

  procedure HideCoulmns;
  var
    j: Integer;
  begin
    for j:=0 to btvItems.ColumnCount - 1 do
      if (btvItems.Columns[j].Position.Band = nil) and btvItems.Columns[j].Visible then
        btvItems.Columns[j].Visible := False;
  end;

  procedure AddBands(AStringList: TStrings);
  var
    i: Integer;
  begin
    for i:=0 to btvItems.Bands.Count - 1 do
      AStringList.Add(btvItems.Bands[i].Caption);
  end;

  procedure RemoveBands(AListBox: TcxListBox);
  var
    i: Integer;
  begin
    for i:=0 to AListBox.Items.Count - 1 do
      if AListBox.Selected[i] then
        DeleteBand(AListBox.Items[i]);
    HideCoulmns;
  end;

begin
  with TViewBandeDemoBandsForm.Create(nil) do
  try
    lbBands.Items.Clear;
    AddBands(lbBands.Items);
    if ShowModal = mrOK then
      RemoveBands(lbBands);
  finally
    Free;
  end;
  UpdateButtons;
end;

procedure TViewBandedDemoMainForm.miShowNavigatorClick(Sender: TObject);
begin
  btvItems.Navigator.Visible := GetMenuItemChecked(Sender);
end;

end.

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/ViewBandedDemo/ViewBandedDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "ViewBandedDemoMain.h"
#include "ViewBandedDemoData.h"
#include "ViewBandeDemoBands.h"
#include "controls.hpp"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridBandedTableView"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBBandedTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxListBox"
#pragma link "cxLookAndFeels"
#pragma link "cxCalendar"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxDBLookupComboBox"
#pragma link "cxGridCardView"
#pragma link "cxImageComboBox"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TViewBandedDemoMainForm *ViewBandedDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TViewBandedDemoMainForm::TViewBandedDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TViewBandedDemoMainForm::DeleteBand(String ABandCaption)
{
  TcxGridBand* Band;
  Band = GetBandByCaption(ABandCaption);
  if(Band != NULL)
    Band->Free();
}

TcxGridBand* __fastcall TViewBandedDemoMainForm::GetBandByCaption(String ABandCaption)
{
  for(int i=0; i < btvItems->Bands->Count; i++)
    if (btvItems->Bands->Items[i]->Caption == ABandCaption){
      return(btvItems->Bands->Items[i]);
    }
  return(NULL);
}

void __fastcall TViewBandedDemoMainForm::FormShow(TObject *Sender)
{
  ((TcxGridTableView*)cxGrid->FocusedView)->DataController->Groups->FullExpand();
}

void __fastcall TViewBandedDemoMainForm::miCreateBandClick(
      TObject *Sender)
{
  String BandCaption;
  if(InputQuery("Create band", "Specify a caption of the band", BandCaption))
    if(GetBandByCaption(BandCaption) != NULL)
       MessageDlg("Band with this caption already exists", mtWarning, TMsgDlgButtons() << mbOK, 0);
    else{
      TcxGridBand *ABand = btvItems->Bands->Add();
      ABand->Caption = BandCaption;
      TcxGridBandViewInfo *ABandViewInfo = 
        btvItems->ViewInfo->HeaderViewInfo->BandsViewInfo->Items[ABand->VisibleIndex];
      btvItems->Controller->LeftPos = ABandViewInfo->Bounds.Right;
    }  
}

void __fastcall TViewBandedDemoMainForm::HideCoulmns()
{
  for(int j=0; j < btvItems->ColumnCount; j++)
    if((btvItems->Columns[j]->Position->Band == NULL) && (btvItems->Columns[j]->Visible))
      btvItems->Columns[j]->Visible = false;
}

void __fastcall TViewBandedDemoMainForm::RemoveBands(TcxListBox* AListBox)
{
  for(int i=0; i < AListBox->Items->Count; i++)
    if(AListBox->Selected[i])
      DeleteBand(AListBox->Items->Strings[i]);
  HideCoulmns();
}

void __fastcall TViewBandedDemoMainForm::AddBands(TStrings* AStringList)
{
  for(int i=0; i < btvItems->Bands->Count; i++)
    AStringList->Add(btvItems->Bands->Items[i]->Caption);
}

void __fastcall TViewBandedDemoMainForm::miDeleteBandClick(TObject *Sender)
{
  TViewBandeDemoBandsForm *ViewBandeDemoBandsForm;
  ViewBandeDemoBandsForm = new TViewBandeDemoBandsForm(NULL);
  try {
    ViewBandeDemoBandsForm->lbBands->Items->Clear();
    AddBands(ViewBandeDemoBandsForm->lbBands->Items);
    if(ViewBandeDemoBandsForm->ShowModal() == mrOk)
      RemoveBands(ViewBandeDemoBandsForm->lbBands);
  }
  __finally {
    ViewBandeDemoBandsForm->Free();
  }
}
//---------------------------------------------------------------------------

void __fastcall TViewBandedDemoMainForm::miColumnsCustomizationClick(TObject *Sender)
{
  btvItems->Controller->Customization = true;
}
//---------------------------------------------------------------------------

void __fastcall TViewBandedDemoMainForm::miShowPreviewRowClick(TObject *Sender)
{
  if (btvItems->Preview->Visible) {
    btvItems->Preview->Visible = false;
    btvItems->Preview->Column = NULL;
  }
  else {
    btvItemsDESCRIPTION->Visible = false;
    btvItems->Preview->Column = btvItemsDESCRIPTION;
    btvItems->Preview->Visible = true;
  };
  ((TMenuItem*)Sender)->Checked = btvItems->Preview->Visible;
}
//---------------------------------------------------------------------------

void __fastcall TViewBandedDemoMainForm::miBandMovingClick(TObject *Sender)
{
  btvItems->OptionsCustomize->BandMoving = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TViewBandedDemoMainForm::miBandSizingClick(TObject *Sender)
{
  btvItems->OptionsCustomize->BandSizing = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TViewBandedDemoMainForm::miShowNavigatorClick(TObject *Sender)
{
  btvItems->Navigator->Visible = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

